### ISSUE TITLE
How to have PlantRunner working on an other strategy

### ISSUE THREAD

**Issue by kunstler** [traitecoevo/plant#187]
I'm trying to develop a new strategy using a different photosynthesis model here https://github.com/kunstler/plantG where I created a new strategy FF16FvCB

I'm not sure why but the function grow_plant_to_size or grow_plant_to_height do not work with the new strategy FF16FvCB.

This is coming from PlantRunner
```
Error in PlantRunner__ctor(plant, environment) : 
  Expected an object of type R6 / PlantPlus<FF16>
```

I think this is coming from https://github.com/traitecoevo/plant/blob/265216a65d6a051bbde184f2f7ab7cf573747d6c/inst/include/plant/plant_runner.h
which is expecting an argument `PlantPlus<FF16_Strategy> plant_` 

This is also specified in https://github.com/traitecoevo/plant/blob/265216a65d6a051bbde184f2f7ab7cf573747d6c/inst/RcppR6_classes.yml
see
```yml

PlantRunner:
  name_cpp: "plant::tools::PlantRunner"
  constructor:
    args: [plant: "plant::PlantPlus<plant::FF16_Strategy>", environment: "plant::Environment"]
  active:
    plant: {access: field, type: "plant::PlantPlus<plant::FF16_Strategy>"}
```

I have try to change that for a template to accept both FF16_Strategy FF16r_Strategy FF16FvCB_Strategy but this is not working ...

Here is what I try 

In  plant/inst/include/plant/plant_runner.h

```cpp
// -*-c++-*-
#ifndef PLANT_PLANT_PLANT_RUNNER_H_
#define PLANT_PLANT_PLANT_RUNNER_H_

#include <plant/plant_plus.h>
#include <plant/environment.h>

namespace plant {
namespace tools {
template <typename T>
// TODO: This should be templated, I think, but that plays badly with
// RcppR6's requirements for "concrete" types.
struct PlantRunner {
// WHAT SHOULD WE DO HERE TO HAVE TWO DIFFERENT PARAMETERS ??
  PlantRunner(PlantPlus<T> plant_, Environment environment_)
    : plant(plant_), environment(environment_) {
    plant.compute_vars_phys(environment);
  }
  static size_t ode_size() {return PlantPlus<T>::ode_size();}
  double ode_time() const {return environment.time;}
  ode::const_iterator set_ode_state(ode::const_iterator it, double time) {
    it = plant.set_ode_state(it);
    environment.time = time;
    plant.compute_vars_phys(environment);
    return it;
  }
  ode::iterator ode_state(ode::iterator it) const {
    return plant.ode_state(it);
  }
  ode::iterator ode_rates(ode::iterator it) const {
    return plant.ode_rates(it);
  }
// WHAT SHOULD WE DO HERE TO HAVE TWO DIFFERENT PARAMETERS ??
  PlantPlus<T> plant;
  Environment environment;
};

}
}

#endif
```
and in RcppR6_classes.yml

```yml
PlantRunner:
  name_cpp: "plant::tools::PlantRunner"
  templates:
    parameters: T
    concrete:
      - ["FF16": "plant::FF16_Strategy"]
      - ["FF16r": "plant::FF16r_Strategy"]
      - ["FF16FvCB": "plant::FF16FvCB_Strategy"]
  constructor:
    args: [plant: "plant::PlantPlus<T>", environment: "plant::Environment"]
  active:
    plant: {access: field, type: "plant::PlantPlus<T>"}
```

But this is not working. I'm not an expert in templating and RcppR6 so @richfitz and @dfalster any help would be appreciate.

**Comment by dfalster** [comment]
Hmm, this is tricky! As noted the `PlantRunner` code is not setup to be templated over different strategy types:
```
// TODO: This should be templated, I think, but that plays badly with
// RcppR6's requirements for "concrete" types.
``
The other templated code (e.g. cohorts, plantPlus, etc) all use class objects, whereas `PlantRunner` is set up as a `stuct`.  This difference might explain the cryptic comment about requirements for concrete types? So the likely way forward here is 

So the way forward is to emulate the setup of other templated objects using a class

**Comment by johnWilshire** [comment]
I feel like I might be close but I could use some help.
I have converted  PlantRunner into a `class` instead of a `struct` looking at `Cohort` and `PlantPlus`as 'guides'. 

I am getting the error when I run `make clean all test`
```
../inst/include/plant/RcppR6_pre.hpp:32:74: error: use of undeclared identifier 'T'
template <> SEXP wrap(const plant::ode::Runner<plant::tools::PlantRunner<T> >&);
```
I think this is because of the OdeRunner in `RccpR6_classes.yml` which I updated to: 

```yml
OdeRunner:
  name_cpp: "plant::ode::Runner<T>"
  templates:
    parameters: T
    concrete:
      - ["Lorenz": "plant::ode::test::Lorenz"]
      - ["OdeR":   "plant::ode::test::OdeR"]
      - ["PlantRunner": "plant::tools::PlantRunner<T>"]
  constructor:
    args: [obj: T, control: "plant::ode::OdeControl=OdeControl()"]
``` 

Note the now required strategy type T in the concrete list thing but am not sure how to continue.
A way forward might be to have OdeRunner be a type with two templated parameters (T and U) as in the RccpR6 "templates" vignette, but I am not sure how that would work  and wanted some guidance before trying this. 

---

For reference: 

My PlantRunner part of `RcppR6_classes.yml`:

```yml
PlantRunner:
  name_cpp: "plant::tools::PlantRunner<T>"
  templates:
    parameters: T
    concrete:
      - ["FF16": "plant::FF16_Strategy"]
      - ["FF16r": "plant::FF16r_Strategy"]
  constructor:
    name_cpp: "plant::tools::PlantRunner<T>"
    args: [plant: "plant::PlantPlus<T>", environment: "plant::Environment"]
  active:
    plant: {access: field, type: "plant::PlantPlus<T>"}
```

My `src/plant_runner.h` :

```cpp
// -*-c++-*-
#ifndef PLANT_PLANT_PLANT_RUNNER_H_
#define PLANT_PLANT_PLANT_RUNNER_H_

#include <plant/plant_plus.h>
#include <plant/environment.h>

namespace plant {
namespace tools {

template <typename T>
class PlantRunner {
public:
  typedef T strategy_type;
  PlantRunner(PlantPlus<T> plant_, Environment environment_)
    : plant(plant_), environment(environment_) {
    plant.compute_vars_phys(environment);
  }

  static size_t ode_size() {return PlantPlus<T>::ode_size();}
  
  double ode_time() const {return environment.time;}
  ode::const_iterator set_ode_state(ode::const_iterator it, double time) {
    it = plant.set_ode_state(it);
    environment.time = time;
    plant.compute_vars_phys(environment);
    return it;
  }
  ode::iterator ode_state(ode::iterator it) const {
    return plant.ode_state(it);
  }
  ode::iterator ode_rates(ode::iterator it) const {
    return plant.ode_rates(it);
  }
  
  PlantPlus<T> plant;
  Environment environment;
};

}
}

#endif
```

**Comment by dfalster** [comment]
Hi @johnWilshire,

Your suggested approach seems at first look right to me. But I also didn't program the templated stuff,  so cannot provide very insightful guidance. I can see it quickly get's complicated, especially with Rcpp code generation also involved. Is it worth trying to get a pure c++ toy example going first?

**Comment by johnWilshire** [comment]
@dfalster  It might help but we would still have to get it back to Rcpp through RcppR6 eventually right?

I am trying something like this instead:
```yml
OdeRunner:
  name_cpp: "plant::ode::Runner<T>"
  templates:
    parameters: T
    concrete:
      - ["Lorenz": "plant::ode::test::Lorenz"]
      - ["OdeR":   "plant::ode::test::OdeR"]
      - ["FF16": "plant::tools::PlantRunner<plant::FF16_Strategy>"]
      - ["FF16r": "plant::tools::PlantRunner<plant::FF16r_Strategy>"]
```
And update when I change the argument type `oderunner_plant_size` in `src/plant_tools.cpp` from just PlantRunner to:

```cpp
plant::PlantPlus_internals
oderunner_plant_size(const plant::ode::Runner<plant::tools::PlantRunner<plant::FF16_Strategy>>& obj) {
  return obj.obj.plant.r_internals();
}

```
temporarily for now, it compiles (yay 😄 ). `make clean all test`

I still need to:

* update code that calls PlantRunner to now call PlantRunner<STRATEGY_NAME_TYPE>
* generalise oderunner_plant_size(...) 
* update the tests / try to write more tests
* test
* update the scaffolder to add this change 

What do you think?

**Comment by dfalster** [comment]
Does the class work if you hardcode in T as FF16?

**Comment by dfalster** [comment]
Seems very promising, well done!

**Comment by johnWilshire** [comment]
Which class do you mean?

`oderunner_plant_size` seems to but its hard to know realy without testing it.

Aside: Because `oderunner_plant_size` just returns the `r_internals()` does it actually need the strategy type dependence?

**Comment by dfalster** [comment]
Sorry, ignore that comment. I posted it before your updated report above came through. Hence i deleted the comment.

**Comment by dfalster** [comment]
Wow, nice work in 381e08b0c8e641455769456e4a5690d4074b3095 @johnWilshire. In particular:

- this was a real challenge, with two layers of templating (ODE plus strategies), and also code generation involved (RcppR6, Rcpp). 
- addition of the `name` function for strategies seems very sensible
- I really appreciate the detailed commit message.
- great that you wrote some extra tests

At first I was taken aback by the amount of code changing, but, in fact, most of the changes were in the generated code (Rcpp exports etc). 

I'll check out those 4 tests to see if I can discover why they don't work.

**Comment by dfalster** [comment]
Before merging this branch, can you also look to

- update the scaffolder
- write some notes the developer notes about the way the templating works.

**Comment by johnWilshire** [comment]
For looking at the tests it might be because `a_f2 = 50` in ff16.h but in `ff16r.h` `a_f2 = 2`.  

I have updated the developer notes but I am not really sure what to say.

**Comment by dfalster** [comment]
Problem with the tests was solved in f9f882f

**Comment by dfalster** [comment]
Closed via PR #191
